### PR TITLE
[New Feature] MOPD

### PR DESCRIPTION
**Pull request by doctorMcy**
# PR type
- [ ] Bug Fix
- [ √ ] New Feature
- [ ] Document Updates
- [ ] More Models or Datasets Support

# PR information
https://github.com/XiaomiMiMo/MiMo-V2-Flash/blob/main/paper.pdf
小米团队推出的MiMo-V2中提出了一种新的后训练范式Multi-Teacher On-Policy Distillation (MOPD)，用于解决能力不平衡的问题（例如：提升了数学，代码能力下降）。
Write the detail information belongs to this PR.

## Experiment results

Paste your experiment result here(if needed).

### CODE REVIEW

File: swift/rlhf_trainers/gkd_trainer.py [hunk 1]
```diff
@@ -298,150 +348,380 @@ def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=N
...
+                            try:
+                                unwrapped = self.accelerator.unwrap_model(teacher_model)
+                                unwrapped_info = f"Unwrapped model type: {type(unwrapped)}, "
+                                if hasattr(unwrapped, 'config'):
+                                    unwrapped_info += f"Unwrapped config attributes: {[attr for attr in dir(unwrapped.config) if not attr.startswith('_')]}"
+                            except Exception as e:
+                                unwrapped_info = f"Failed to unwrap model: {e}"
+
+                            raise ValueError(f"Cannot determine teacher model path for tokenizer initialization. "
+                                             f"Model info: {model_info}. Unwrapped info: {unwrapped_info}")
```
Comment (tpx818) [inline swift/rlhf_trainers/gkd_trainer.py]: 这一段可以抽象一个函数

**Comment by Tohrusky** [comment]
很乐意帮忙测试 有合并的计划吗~ @Jintao-Huang

**Comment by Tohrusky** [comment]
image

I think it would be better if adding a shell in the examples directory to show how to start training. 

Specifically, it'll show how to launch the vLLM rollout with multiple teachers, start MOPD training, and whether use reward functions(ORM) as a part of A in the paper.

**Comment by Jintao-Huang** [comment]
cc @hjh0119 

> 很乐意帮忙测试 有合并的计划吗~ @Jintao-Huang

**Comment by hjh0119** [comment]
Thanks for the contribution, feel free to let me know when it's ready to merge

File: swift/arguments/rlhf_args.py [hunk 2]
```diff
@@ -65,6 +65,9 @@ class TeacherModelArguments:
             remotely. When this is set, `teacher_model` is not required. Defaults to None.
     """
     teacher_model: Optional[str] = None
+    teacher_model_group: List[str] = field(default_factory=list)
```
Comment (hjh0119) [inline swift/arguments/rlhf_args.py]: Consider changing teacher_model to Optional[List[str]] (similar to reward_model) to avoid introducing additional parameters

File: swift/pipelines/train/rlhf.py [hunk 3]
```diff
@@ -132,6 +132,24 @@ def _prepare_model_tokenizer(self):
...
+                # Use teacher_model_type and teacher_model_revision if available, otherwise infer
+                model_type = getattr(args, 'teacher_model_type', None)
+                model_revision = getattr(args, 'teacher_model_revision', None)
+
+                result = self._prepare_single_model_for_teacher_group(teacher_model_path, model_type, model_revision)
+                if result is not None:
+                    model, _ = result
+                    self.teacher_model_group_models.append(model)
+                    logger.info(f'Successfully loaded teacher model group [{idx}]: {model}')
+            logger.info(f'Total teacher_model_group_models loaded: {len(self.teacher_model_group_models)}')
```
Comment (hjh0119) [inline swift/pipelines/train/rlhf.py]: Similarly, just like the handling of reward_model, without introducing additional logic

File: swift/pipelines/train/rlhf.py [hunk 4]
```diff
@@ -166,6 +184,44 @@ def _prepare_model_tokenizer(self):
...
+                num_labels=num_labels)
+
+        # For teacher models, set to eval mode and disable gradients
+        if self.args.sequence_parallel_size > 1:
+            sequence_parallel.prepare(
+                self.args.sequence_parallel_size, model, processor, padding_free=args.padding_free)
+        model.requires_grad_(False).eval()
+
+        HfConfigFactory.set_config_attr(model.config, 'use_cache', False)
+        return model, processor
```
Comment (hjh0119) [inline swift/pipelines/train/rlhf.py]: Same as above

File: swift/pipelines/train/rlhf.py [hunk 5]
```diff
@@ -238,7 +294,16 @@ def _get_trainer_kwargs(self):
...
+            # Pass pre-loaded teacher_model_group_models if available, otherwise pass the string list
+            if hasattr(self, 'teacher_model_group_models') and self.teacher_model_group_models:
+                trainer_kwargs['teacher_model_group_models'] = self.teacher_model_group_models
+            else:
+                trainer_kwargs['teacher_model_group'] = self.args.teacher_model_group
             trainer_kwargs['teacher_use_disable_adapter'] = getattr(self.args, '_teacher_use_disable_adapter', False)
+            if self.args.use_mopd:
+                trainer_kwargs['use_mopd'] = self.args.use_mopd
+                # todo
+                # trainer_kwargs['mopd_config'] = self.args.mopd_config
```
Comment (hjh0119) [inline swift/pipelines/train/rlhf.py]: Same as above

File: swift/rlhf_trainers/gkd_trainer.py [hunk 6]
```diff
@@ -133,6 +138,60 @@ def __init__(self, model: Optional[Union[PreTrainedModel, nn.Module, str]] = Non
...
+                                'use_extended_uld': True,
+                                'uld_use_hybrid_loss': True,
+                                'uld_crossentropy_weight': 0.0,
+                                'uld_distillation_weight': 1.0,
+                                'uld_student_temperature': 1.0,
+                                'uld_teacher_temperature': 1.0,
+                            },
+                            student_tokenizer=self.student_tokenizer,
+                            teacher_tokenizer=teacher_tokenizer,
+                        )
```
Comment (hjh0119) [inline swift/rlhf_trainers/gkd_trainer.py]: For now, without considering the cross-tokenizer case, don't introduce GOLDLoss as it adds unnecessary complexity

File: swift/rlhf_trainers/gkd_trainer.py [hunk 7]
```diff
@@ -169,7 +228,7 @@ def get_train_dataloader(self):
     def _build_opsd_teacher_data(self, inputs):
         """Build teacher data for OPSD by replacing the last user message with teacher_prompt.
 
-        Returns None if teacher_prompt is not available in all examples.
+        Returns None if teacher_prompt is not av ailable in all examples.
```
Comment (hjh0119) [inline swift/rlhf_trainers/gkd_trainer.py]: restore it

File: swift/rlhf_trainers/gkd_trainer.py [hunk 8]
```diff
@@ -299,7 +358,33 @@ def generate_on_policy_outputs(self, model, inputs, generation_config, pad_token
         new_position_ids = new_attention_mask.cumsum(dim=1) - 1
         new_position_ids[new_position_ids < 0] = 0
         inputs['position_ids'] = new_position_ids
-        return generated_tokens, new_attention_mask, new_labels
+        if self.use_mopd:
+            # 返回解码后的文本encoded_inputs
```
Comment (hjh0119) [inline swift/rlhf_trainers/gkd_trainer.py]: Please write comments in English only

File: swift/rlhf_trainers/gkd_trainer.py [hunk 9]
```diff
@@ -299,7 +358,33 @@ def generate_on_policy_outputs(self, model, inputs, generation_config, pad_token
...
+
+                # 解码清理后的 token IDs
+                completion_text = self.template.safe_decode(cleaned_ids)
+
+                completion_text = completion_text.strip()
+
+                completion_texts.append(completion_text)
+        else:
+            completion_texts = None
+        return generated_tokens, new_attention_mask, new_labels, completion_texts
```
Comment (hjh0119) [inline swift/rlhf_trainers/gkd_trainer.py]: Why do we need to parse completion_text here?

File: swift/rlhf_trainers/gkd_trainer.py [hunk 10]
```diff
@@ -447,6 +532,71 @@ def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=N
...
+                loss_total = gold_adapter(
+                    student_logits=outputs_student.logits,
+                    teacher_logits=teacher_logits,
+                    student_labels=student_labels,
+                    teacher_labels=teacher_labels,
+                    student_input_ids=student_input_ids,
+                    teacher_input_ids=teacher_input_ids,
+                )
+                loss += loss_total / len(teacher_model_group)
+        # Separate teacher model provided
```
Comment (hjh0119) [inline swift/rlhf_trainers/gkd_trainer.py]: Same as above, please update the multi-teacher logic based on the existing single-teacher logic

Comment (hjh0119) [inline swift/rlhf_trainers/gkd_trainer.py]: no return value here

File: swift/rlhf_trainers/gkd_trainer.py [hunk 11]
```diff
@@ -470,11 +620,114 @@ def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=N
...
+        # Additional fallback: try to get from unwrapped model's name_or_path
+        if model_path is None:
+            unwrapped_student = self.accelerator.unwrap_model(model)
+            model_path = getattr(unwrapped_student, 'name_or_path', None)
+        # Additional fallback: try to get from unwrapped model's config name_or_path
+        if model_path is None:
+            unwrapped_student = self.accelerator.unwrap_model(model)
+            if hasattr(unwrapped_student, 'config') and hasattr(unwrapped_student.config, 'name_or_path'):
+                model_path = unwrapped_student.config.name_or_path
+        return model_path
```
Comment (hjh0119) [inline swift/rlhf_trainers/gkd_trainer.py]: remove, reuse origin logic

File: swift/rlhf_trainers/gkd_trainer.py [hunk 12]
```diff
@@ -913,16 +1166,16 @@ def _align_vocab_size(student_logits, teacher_logits):
...
+            self,
+            student_logits,
+            teacher_logits=None,
+            labels=None,
+            beta=0.5,
+            temperature=1.0,
+            chunk_size=512,
+            topk=None,
+            teacher_topk_logprobs=None,
+            teacher_topk_indices=None,
```
Comment (hjh0119) [inline swift/rlhf_trainers/gkd_trainer.py:980]: restore it

File: swift/rlhf_trainers/gold_loss_adapter.py [hunk 13]
```diff
@@ -0,0 +1,524 @@
...
+        # 使用log_target=True，传入log概率
+        kl_p_m = F.kl_div(log_m, log_student, reduction='batchmean', log_target=True)
+        kl_q_m = F.kl_div(log_m, log_teacher, reduction='batchmean', log_target=True)
+        jsd = 0.5 * (kl_p_m + kl_q_m)
+
+        # 检查结果是否有效
+        if torch.isnan(jsd) or torch.isinf(jsd):
+            return torch.tensor(0.0, device=student_probs.device, requires_grad=True)
+
+        return jsd
```
Comment (hjh0119) [inline swift/rlhf_trainers/gold_loss_adapter.py]: Same as above, without considering GOLDLoss, limit MOPD to the premise of using the same tokenizer

File: swift/arguments/rlhf_args.py (hunk 2 above)
Comment (hjh0119) [reply swift/arguments/rlhf_args.py]: No need for an extra use_mopd parameter, MOPD can be determined by the number of teacher models

**Review by hjh0119 — CHANGES_REQUESTED** [review]

**Comment by hjh0119** [comment]
Please keep the PR clean. For MOPD, I think we can reuse the existing GKD pipeline and simply extend the original teacher_model related logic to support multiple teachers

**Comment by Kagura-0001** [comment]
加油

**Comment by doctorMcy** [comment]
> Please keep the PR clean. For MOPD, I think we can reuse the existing GKD pipeline and simply extend the original teacher_model related logic to support multiple teachers

感谢您的审阅。我希望能支持不同tokenizer的模型之间进行蒸馏，是否单独写一个trainer类比较好？

**Comment by hjh0119** [comment]
I would advise against supporting cross-tokenizer distillation for a couple of reasons:

1. To my knowledge, current frontier model research and academic work primarily focus on distillation between models of the same family (i.e., same tokenizer).
2. Introducing cross-tokenizer support would add significant code complexity, regardless of whether it is integrated into the existing `GKDTrainer` or implemented as a separate Trainer.

The current `GKDTrainer` has already accumulated a considerable number of features. In the next iteration, I plan to refactor it to simplify the codebase and explore a design that supports multi-teacher distillation scenarios. As such, I would recommend holding off on further development in this direction for now.

That said, thank you for your contribution — it is very much appreciated!

**Comment by mccatec** [comment]
> I would advise against supporting cross-tokenizer distillation for a couple of reasons:
> 
> 1. To my knowledge, current frontier model research and academic work primarily focus on distillation between models of the same family (i.e., same tokenizer).
> 2. Introducing cross-tokenizer support would add significant code complexity, regardless of whether it is integrated into the existing `GKDTrainer` or implemented as a separate Trainer.
> 
> The current `GKDTrainer` has already accumulated a considerable number of features. In the next iteration, I plan to refactor it to simplify the codebase and explore a design that supports multi-teacher distillation scenarios. As such, I would recommend holding off on further development in this direction for now.
> 
> That said, thank you for your contribution — it is very much appreciated!

Agreed, though there are algorithms such as GOLD that support cross-tokenizer distillation. However, it is essentially just log-probability alignment, that is, making the log-probability distributions as similar as possible. This kind of distribution alignment is not syntax-related, so it brings little gain after distillation. We tested cross-tokenizer distillation from Gemma 4 to Qwen on the IF dataset, and IFEval showed almost the same performance.

**Comment by hjh0119** [comment]
multi-teacher has been supported in https://github.com/modelscope/ms-swift/pull/9678
